### PR TITLE
FSStore: use `ensure_bytes()`

### PR DESCRIPTION
**Pull request by madsbk**
Fixes https://github.com/zarr-developers/zarr-python/issues/1279

TODO:
* [x] Add unit tests and/or doctests in docstrings
* [x] Add docstrings and API docs for any new/modified user-facing classes and functions
* [ ] Changes documented in docs/release.rst
* [ ] GitHub Actions have all passed
* [ ] Test coverage is 100% (Codecov passes)

### CODE REVIEW

**Comment by madsbk** [comment]
@normanrz can you check if this fixes the issue?

**Comment by normanrz** [comment]
Yes! That looks good to me. Thanks for the quick fix!

**Comment by rabernat** [comment]
> We changed this with #934, now we also allow [`NDArrayLike`](https://github.com/zarr-developers/numcodecs/blob/922ab23d85a9f854352782212a5e057b7e31e754/numcodecs/ndarray_like.py#L45) instances.

I see that this has been implemented in the zarr-python implementation. And I thank everyone who worked on this amazing feature! 🚀  But that doesn't mean it is consistent with what the [spec says](https://zarr-specs.readthedocs.io/en/latest/core/v3.0.html#abstract-store-interface):

> The store interface defines a set of operations involving keys and values. In the context of this interface, a key is any string... A value is a sequence of bytes.

It is certainly straightforward to convert an uncompressed ndarray into a sequence of bytes. So maybe I am overthinking this... Or maybe the spec needs to be updated to reflect these new implementation changes? 🤔  My only goal here is to make sure that the spec and the implementation are consistent with each other. Otherwise we risk interoperability with other zarr implementations.

If we are fine with saying "a value is a sequence of bytes or a raw array" then I am fine with leaving it up to the store implementation whether to convert arrays to bytes.

**Comment by madsbk** [comment]
Good point and I agree, the spec and implementation should match up!

I must admit I read "A value is a sequence of bytes" in a more general sense, not  a reference to the `bytes` class specifically. Otherwise,  types like `bytearray` or `memoryview` are also excluded.

**Comment by jakirkham** [comment]
Hey Mads, thanks for following up on this issue.

That said, don't think we should be making this change. In general it is already the expectation in zarr-python & Numcodecs that [Python Buffer Protocol]( https://docs.python.org/3/c-api/buffer.html ) supporting objects are used (`memoryview`s, `ndarray`s, `bytearray`s, `bytes`, etc.). Part of the reason for this is it allows zero-copy operations on the data (like casting, reshaping, etc.). It also can support things like memory-mapping natively (as `mmap` supports the Python Buffer Protocol).

How to word this in the spec is nuanced. The spec itself is not meant to be Python specific, but to support other languages as well. They may not have `bytearray` or `memoryview` for example or if they do it may be named differently and/or behave differently. So think we should consider language agnostic wording. It sounds like this is how you understood it as well.

Perhaps others who found this wording in the spec confusing can chime in on how to clarify it. Maybe we can say "a value contains binary data"?

**Comment by madsbk** [comment]
@jakirkham, I am not sure I understand. Are you saying that we shouldn't use `ensure_bytes()` in `FSStore` (this PR) or that the spec should stay language agnostic? 

I have no strong opinion about the specifications, but I think it is important that we allow  `NDArrayLike` objects in `values`. If we move `ensure_bytes()` to the `zarr.core` level, we are effectively saying that a `value` is always of type `bytes`.

**Comment by jakirkham** [comment]
Could we call `ensure_ndarray_like` instead?

**Comment by normanrz** [comment]
> Could we call `ensure_ndarray_like` instead?

I'm not sure that would work. I think the trick is that at some point, `tobytes` needs to be called with `value.tobytes(order='A')` to preserve the order. https://numpy.org/doc/stable/reference/generated/numpy.ndarray.tobytes.html

**Comment by jakirkham** [comment]
There's `ensure_contiguous_ndarray_like`, which does that

**Comment by jstriebel** [comment]
Since most other stores use `ensure_bytes` atm, I think this fix is valid for the moment, and important to fix an existing bug. I agree, that there should be a discussion whether every store should do this or not, but this should not block this PR IMO.

**Review by rabernat — APPROVED** [review]
I am fine with this fix for now, since it seems consistent with what most other stores do.

We can revisit the question of whether this `ensure_bytes` logic should live elsewhere in the future.

**Comment by jakirkham** [comment]
Sorry to be a stickler. The main reason I bring up `ensure_bytes` here is this typically incurs the cost of a copy (with additional memory being used). Given this is happening with multiple memory segments that are being written together, this could result in more significant memory usage. Especially as we start to rely more on batching multiple reads & writes. While it may not always be possible to avoid these copies, they could be deferred until absolutely necessary (with the resulting `bytes` objects quickly discarded) to keep memory usage as low as possible (since not all memory segments would needing copying at once). Hope that clarifies the concern a bit more.

File: zarr/storage.py
Comment (jakirkham) [inline zarr/storage.py:1384]: Maybe something like this would work

```suggestion
        norm_values = {}
        for key, val in values.items():
            norm_key = self._normalize_key(key)
            try:
                val = ensure_contiguous_ndarray_like(val)
            except TypeError:
                val = ensure_bytes(val)
            norm_values[norm_key] = val
        self.map.setitems(norm_values)
```

**Comment by normanrz** [comment]
Any chance we could get this merged and released?

**Comment by rabernat** [comment]
I'd like to hear from @jakirkham - John, would you be you okay with addressing  `ensure_contiguous_ndarray_like` vs. `ensure_bytes` in a follow-on PR (where we can also address it for all the other stores that call `ensure_bytes`)? Or do you think this needs to be resolved before we merge?

**Comment by normanrz** [comment]
I think @jakirkham's [latest comment](https://github.com/zarr-developers/zarr-python/pull/1285#discussion_r1048880317) is a straightforward compromise that could be incorporated as part of this PR.

**Comment by madsbk** [comment]
> I think @jakirkham's [latest suggestion](https://github.com/zarr-developers/zarr-python/pull/1285#discussion_r1048880317) is a straightforward compromise that could be incorporated as part of this PR.

Done

**Comment by joshmoore** [comment]
Re-opening to try to fix the codecov build.

**Comment by joshmoore** [comment]
Now green. @jakirkham: does this match your expectations?

If so, happy to get it in and prep a release during the holidays. (Help with the release notes will be appreciated!)

**Comment by rabernat** [comment]
If we're doing a release, let's get #1304 in too.

**Comment by joshmoore** [comment]
> If we're doing a release, let's get #1304 in too.

see https://github.com/zarr-developers/zarr-python/pull/1309#issuecomment-1384179684

Leaving this PR for @jakirkham to review.

File: zarr/util.py
Comment (jakirkham) [inline zarr/util.py:706]: We may want to move this to [`numcodecs.compat`]( https://github.com/zarr-developers/numcodecs/blob/67ede4c6b4f1707f1da18351945e10904a5572de/numcodecs/compat.py ), but that can be follow-on work

File: zarr/storage.py
Comment (jakirkham) [inline zarr/storage.py:1403]: Something we may consider is delaying the instantiation of the `dict` so these copies occur as these values are requested. This can be a follow-on though

**Review by jakirkham — APPROVED** [review]
Thanks Mads! 🙏

LGTM. Made a few comments that we can follow up on separately

**Comment by joshmoore** [comment]
@jakirkham: ok to get this in the 2.13.x series?

@madsbk: I'll get this out ASAP but might wait on any emergency needs for re: #1324

Suggestions for the docs/release.rst statement would be welcome. 🙏🏽

**Comment by jakirkham** [comment]
> Suggestions for the docs/release.rst statement would be welcome. 🙏🏽

Oops sorry should have asked for that here (or just added it). Submitted PR ( https://github.com/zarr-developers/zarr-python/pull/1325 )

**Pull request merged**
